firebase-storage/src/main/java/com/google/firebase/storage/StorageRegistrar.java

```diff
@@ -16,19 +16,31 @@
 
 import android.support.annotation.Keep;
 import android.support.annotation.RestrictTo;
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.auth.internal.InternalAuthProvider;
 import com.google.firebase.components.Component;
 import com.google.firebase.components.ComponentRegistrar;
+import com.google.firebase.components.Dependency;
 import com.google.firebase.platforminfo.LibraryVersionComponent;
-import java.util.Collections;
+import java.util.Arrays;
 import java.util.List;
 
 /** @hide */
 @Keep
 @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
 public class StorageRegistrar implements ComponentRegistrar {
   @Override
+  @Keep
   public List<Component<?>> getComponents() {
-    return Collections.singletonList(
+    return Arrays.asList(
+        Component.builder(FirebaseStorageComponent.class)
+            .add(Dependency.required(FirebaseApp.class))
+            .add(Dependency.optional(InternalAuthProvider.class))
```

Comment:
With my limited understanding of storage it looks like `InternalAuthProvider` is not needed right away(when storage initializes) unlike in `Firestore`, but only used in response to user calls to downloads and what not. If that is the case consider switching to `Dependency.optionalProvider(InternalAuthProvider.class)`, what it will do is defer initialization of `InternalAuthProvider` until it is needed as well as prevent potential dependency cycles in the future.

That said, currently the only implementor of `InternalAuthProvider` is `FirebaseAuth` which is eager, so deferred initialization will not apply, but it seems a relatively easy change to do(especially if the `Provider<T>` is hidden behind storage.getAuthProvider()) to bulletproof storage for the future. wdyt?

Reply:
Your understanding is correct. We only fetch a token once a user requests a file upload or download, so we don't need the AuthProvider until later. The change for this was relatively minor, so I updated this PR.

BTW, if wasn't immediately obvious to me whether optionalProvider  makes Provider nullable or whether it makes Provider.get() nullable. I think it's the former (from looking through the code and playing with it). Maybe we can document this?